Many minor optimisations for 3.13/3.14

gudnimg opened the pull request:
WIP. I'm adding a few minor optimisations to this branch which we probably don't want to create a separate PR for each since they save little < 10B each.

Current savings:
Flash: -420bytes
SRAM:  -2 bytes

gudnimg committed:
cleanup: remove ctype include

We don't use any function from this header

No change in code size

gudnimg committed:
optimisation: mmu: combine if statements

Change in memory:
Flash: -6 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation: simplify T-code response for single material

Extruder is always 0

Change in memory:
Flash: -12 bytes
SRAM: 0 bytes

gudnimg committed:
optimsation: draw cursor and set position with one function

Change in memory:
Flash: -4 bytes
SRAM: 0 bytes

gudnimg committed:
cleanup: remove unused variable FSensorStateMenu

gudnimg committed:
optimisation: make clock_interval static and 1 byte

Change in memory:
Flash: -16 bytes
SRAM: -1 byte

gudnimg committed:
cleanup: make heating_status_counter static

Change in memory:
Flash: 0 bytes
SRAM: 0 bytes

gudnimg committed:
cleanup: remove unused function declaration

gudnimg committed:
optimisation: use new EEPROM init method for fCheckModeInit

Change in memory:
Flash: -90 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation: utils: don't initialise global variables

They will be set by fCheckModeInit() by reading from EEPROM

Change in memory:
Flash: -2 bytes
SRAM: -1 byte

gudnimg committed:
optimisation: use new eeprom methods in check_printer_version()

Change in memory:
Flash: -48 bytes
SRAM: 0 bytes

gudnimg committed:
M115: Extruder count is always 1

Change in memory:
Flash: -8 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation: more lcd_putc_at

Change in memory:
Flash: -18 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation: lcd_temp_cal_show_result

result is boolean so always update the EEPROM

Change in memory:
Flash: -20 bytes
SRAM: 0 bytes

gudnimg committed:
use lcd_puts_P instead of lcd_printf_P

Change in memory:
Flash: -8 bytes
SRAM: 0 bytes

gudnimg committed:
cleanup: remove unused global variable

gudnimg commented:
Rebased to sync with MK3 branch and fix merge conflict.

Also dropped one questionable optimisation.

gudnimg commented:
Rebased  to sync with MK3 branch and resolve conflicts.

gudnimg committed:
optimisation: combine if statements in Sensors menu

FINDA status is always 0 or 1

Change in memory:
Flash: -28 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation: use expired_cont is specific places

!variable.running() || variable.expired()
is equal to
variable.expired_cont()

The latter required less flash memory

Change in memory:
Flash: -28 bytes
SRAM: 0 bytes

leptun committed:
optimize get_coordinates() axis mask

leptun committed:
Remove dead code from get_coordinates()

leptun committed:
Remove useless parantheses

Found these a few months ago when I was trying some other optimization. Don't want to make a PR just for this, so I'll just slide it here in this existing PR.

leptun commented on Firmware/ultralcd.cpp:
```diff
@@ -3224,17 +3224,16 @@ void lcd_temp_cal_show_result(bool result) {
-	else {
-		eeprom_update_byte((uint8_t*)EEPROM_CALIBRATION_STATUS_PINDA, 0);
+	} else {
 		SERIAL_ECHOLNPGM("PINDA calibration failed. Continue with pressing the knob.");
```
Can this serial message be extracted to a single PROGMEM item? If I recall, these PSTR strings are not deduplicated by the linker script

leptun replied:
Scrap this. Not sure how I didn't see that one string says "done" and one says "failed". So you may ignore the comment above.

leptun commented on Firmware/util.cpp:
```diff
@@ -247,36 +247,15 @@ ClCheckVersion oCheckVersion=ClCheckVersion::_None;
-        oCheckGcode = ClCheckGcode::_Warn;
-        eeprom_update_byte((uint8_t *)EEPROM_CHECK_GCODE, (uint8_t)oCheckGcode);
-    }
+    oCheckMode = (ClCheckMode)eeprom_init_default_byte((uint8_t *)EEPROM_CHECK_MODE, farm_mode ? (uint8_t) ClCheckMode::_Strict : (uint8_t)ClCheckMode::_Warn);
```
I don't think this change is completely correct. We can't rely on farm_mode during eeprom initialization since farm mode will most likely not be enabled during the first eeprom initialization, so the default value would always be loaded instead.
The old code also seems to enforce a few of these settings even if the eeprom is already initialized. The new code only sets defaults when uninitialized. I think we should try to preserve the old behaviour.

The EEPROM_NOZZLE_DIAMETER_uM_DEFAULT initialization looks as if it behaves correctly, but I'm a bit confused by the old code. It's overly complicated and I still don't understand why we need two eeprom "nozzle diameter" settings.

The EEPROM_CHECK_MODEL, EEPROM_CHECK_VERSION and EEPROM_CHECK_GCODE code looks correct 👍.

gudnimg replied:
Regarding `EEPROM_CHECK_MODE`, does the change below look OK? Surprisingly there is no change in memory usage.

The code below will always at least initialise the EEPROM value. But if farm mode is enabled, it's overwritten if it hasn't been done already.
![image](https://user-images.githubusercontent.com/8218499/231295308-f6d8c4ff-9442-467d-9f4c-797f5b6ca49f.png)

leptun replied:
I think that change is correct, but I didn't check it too carefully.
There is still EEPROM_NOZZLE_DIAMETER below it that doesn't look correct. We can't have a ternary operator on farm_mode in one of those eeprom _init_default functions.

gudnimg replied:
I havent forgotten the nozzle setting. Just needed a fresh 🧠 and sleep. I need to re-review how these nozzle eeprom settings are used. It’s very confusing at only a quick glance. Needs a more thorough “glance” ;) 

I may delay this for the weekend to priorities the menu testing. But will see how the week goes. :)

gudnimg replied:
Hopefully resolved with [bc4089f](https://github.com/prusa3d/Prusa-Firmware/pull/4027/commits/bc4089fe6517803a8fb11be7ff33b5a531533ce8)

leptun commented:
Please see the comments I wrote above. Besides those, everything looks good to me 👍.

gudnimg committed:
cleanup: make cancel_heatup static in Marlin_main

gudnimg committed:
cleanup: tmc2130: make variables static where applicable

No change in memory

gudnimg commented:
Rebased to resolve conflict

gudnimg committed:
Move fan_state to fancheck.cpp and make it static

No change in memory

gudnimg committed:
Simplify two fan_check_error checks

There are only 3 possible values. No need to check for two of them
to rule of the last value. Instead simply check for the last value only.

Change in memory:
Flash: -16 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation: lcd_putc_at in belt test

Change in memory:
Flash: -4 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation: belt test new values

use lcd_print instead of lcd_printf_P
to print the number only

Tested on MK3S+

Change in memory:
Flash: -34 bytes
SRAM: 0 bytes

gudnimg committed:
Resolve PR review

Initialise EEPROM_CHECK_MODE always
But if farm mode is detected, update the Check mode to Strict

Initialise EEPROM_NOZZLE_DIAMETER on farm mode.
When opening the menus this value is set automatically to
_Diameter_400 anyway so this should be fine.

No change in memory on Multilang build

gudnimg committed:
many more lcd_putc_at optimisations

Changes in
lcd_implementation_drawmenu_sdfile
and
lcd_implementation_drawmenu_sddirectory
seem to save the most.

Change in memory:
Flash: -72 bytes
SRAM: 0 bytes

gudnimg commented:
Rebased to pull in latest fixes on the MK3 branch. Such as the editing menu fixes.

gudnimg committed:
cleanup: static variables in Marlin_main.cpp

wavexx commented on Firmware/Marlin_main.cpp:
```diff
@@ -183,13 +183,12 @@ bool mesh_bed_leveling_flag = false;
 
 
-bool fan_state[2];
+
```
Would be nice to remove those extra useless lines.

wavexx commented on Firmware/Marlin_main.cpp:
```diff
@@ -6472,10 +6464,9 @@ SERIAL_PROTOCOLPGM("\n\n");
-          SERIAL_ECHOPGM(STRINGIFY(EXTRUDERS)); 
-          SERIAL_ECHOPGM(" UUID:"); 
+          SERIAL_ECHOPGM(CUSTOM_MENDEL_NAME);
+          SERIAL_ECHOPGM(" EXTRUDER_COUNT:1");
```
Since we still have "EXTRUDERS" we can probably do:

``SERIAL_ECHOPGM(" EXTRUDER_COUNT:" STRINGIFY(EXTRUDERS)));``

which should still avoid the extra call.

leptun committed:
Remove empty lines

leptun committed:
Use `EXTRUDERS` instead of hardcoded value in string
